### PR TITLE
[Next.js] Personalize Middleware

### PR DESCRIPTION
**Pull request by ambrauer**
## Description / Motivation
This PR includes productization of the Next.js personalization middleware (based on original POC). Also includes fixes for `CdpIntegrationScript.tsx`.

Notably missing here are unit tests for `PersonalizeMiddleware` (still determining best way to unit test edge middleware) and MSW-based mocks of CDP API endpoints (will be a separate task).

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/sitecore-jss/src/personalize/utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    it('should return the root path with the segment id', () => {
+      const pathname = '/';
+      const result = getPersonalizedRewrite(pathname, data);
+      expect(result).to.equal(`/${SEGMENT_PREFIX}${data.segmentId}/`);
+    });
+  });
+
+  describe('getPersonalizedRewriteData', () => {
+    it('should return a PersonalizedRewriteData object', () => {
+      expect(getPersonalizedRewriteData('/some/path')).to.be.an('object');
```
[the remark this answers is not in the document]
Comment (sc-addypathania) [reply packages/sitecore-jss/src/personalize/utils.test.ts:36]: @illiakovalenko this test is just about checking the return type of the function.
The test below this checks for what you're referring to.

File: packages/sitecore-jss/src/personalize/utils.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+export const DEFAULT_SEGMENT = '_default';
+export const SEGMENT_PREFIX = '_segmentId_';
+
+export type PersonalizedRewriteData = {
+  segmentId: string;
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss/src/personalize/utils.ts:5]: Correct, it's possible there may be more than segmentId embedded in the rewrites in the future.

File: packages/sitecore-jss-nextjs/src/index.ts [hunk 3]
```diff
@@ -52,7 +52,13 @@ export {
   RestDictionaryService,
   RestDictionaryServiceConfig,
 } from '@sitecore-jss/sitecore-jss/i18n';
-export { personalizeLayout } from '@sitecore-jss/sitecore-jss/personalize';
+export {
+  personalizeLayout,
+  PersonalizedRewriteData,
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss-nextjs/src/index.ts:59]: Probably not? Removed.

File: packages/sitecore-jss/src/personalize/graphql-personalize-service.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  private graphQLClient: GraphQLClient;
+
+  protected get query(): string {
+    return /* GraphQL */ `
+      query($siteName: String!, $language: String!, $itemPath: String!) {
+        layout(site: $siteName, routePath: $itemPath, language: $language) {
+          item {
+            id
+            version
+            personalization {
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss/src/personalize/graphql-personalize-service.ts:51]: 😮 that is strange... I hadn't noticed this. Might be worth bringing up to the back-end team.

File: packages/sitecore-jss/src/native-fetcher.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
+import { HttpResponse } from './data-fetcher';
+import debuggers, { Debugger } from './debug';
+
+type NativeDataFetcherOptions = {
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/native-fetcher.ts:4]: We should allow to pass more arguments, b/c we expect that on the [line](https://github.com/Sitecore/jss/pull/1008/files#diff-04aaa9815a34b8a043be4f5d5cf48bd5060f6c2eaf2a2462be70d597c59c46d8R27) using init
```suggestion
type NativeDataFetcherOptions = {
  [key: string]: unknown
```

File: packages/sitecore-jss/src/native-fetcher.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      statusText: response.statusText,
+      headers: response.headers,
+      url: response.url,
+      redirected: response.redirected,
+      data: respData,
+    };
+
+    if (!response.ok) {
+      debug('response error: %o', debugResponse);
+      throw new Error(`HTTP ${response.status} ${response.statusText}`);
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/native-fetcher.ts:54]: See the comment above, probably just wrap everything by `try catch` and throw the error from here, but in the `catch` handle that in the way like here

File: packages/sitecore-jss-nextjs/src/edge/personalize-middleware.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      const expiryDate = new Date(new Date().setFullYear(new Date().getFullYear() + 2));
+      const options = { expires: expiryDate, secure: true };
+      res.cookie(this.browserIdCookieName, browserId, options);
+    }
+  }
+
+  private excludeRoute(pathname: string) {
+    if (
+      pathname.includes('.') || // Ignore files
+      pathname.startsWith('/api') // Ignore API calls
```
Comment (illiakovalenko) [inline packages/sitecore-jss-nextjs/src/edge/personalize-middleware.ts:87]: It's possible that pathname could have a route name that starts with `/api` in this case it will be also excluded, need to do more strict check, using RegExp probably

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
Looks very nice 👍  see some suggestions below

File: packages/sitecore-jss-nextjs/src/edge/redirects-middleware.ts [hunk 8]
```diff
@@ -75,7 +75,7 @@ export class RedirectsMiddleware {
 
   /**
    * Method returns RedirectInfo when matches
-   * @param url
+   * @param {URL} url
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss-nextjs/src/edge/redirects-middleware.ts:78]: Fixed

File: packages/sitecore-jss/src/personalize/graphql-personalize-service.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+   */
+  constructor(protected config: GraphQLPersonalizeServiceConfig) {
+    this.graphQLClient = this.getGraphQLClient();
+  }
+
+  /**
+   * Get personalize information for a route
+   * @param {string} itemPath page route
+   * @param {string} language language
+   * @returns {PersonalizeInfo | undefined} the personalize information or undefined (if itemPath / language not found)
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss/src/personalize/graphql-personalize-service.ts:72]: Fixed

File: packages/sitecore-jss/src/native-fetcher.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    return {
+      ...response,
+      data: respData as T,
+    };
+  }
+
+  /**
+   * Determines settings for the request
+   * @param {RequestInit} init Custom settings for request
+   * @param {unknown} data Optional data to POST with the request
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss/src/native-fetcher.ts:66]: Fixed

File: packages/sitecore-jss/src/native-fetcher.ts (hunk 5 above)
Comment (ambrauer) [reply packages/sitecore-jss/src/native-fetcher.ts:4]: We have this covered with `NativeDataFetcherConfig` (combination of `NativeDataFetcherOptions` and `RequestInit`):
https://github.com/Sitecore/jss/pull/1008/files#diff-04aaa9815a34b8a043be4f5d5cf48bd5060f6c2eaf2a2462be70d597c59c46d8R15

File: packages/sitecore-jss-nextjs/src/edge/personalize-middleware.ts (hunk 7 above)
Comment (ambrauer) [reply packages/sitecore-jss-nextjs/src/edge/personalize-middleware.ts:87]: I updated to `/api/` which should reduce any accidental exclusions of legitimate routes. We provide the [`excludeRoute`](https://github.com/Sitecore/jss/pull/1008/files#diff-c5f8141e244ca57373cab027e38bab2cbf958cefb82c4fbbbfd7779493b8bdb3R19) option for them to customize further if necessary.

File: packages/sitecore-jss/src/native-fetcher.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    // Note a goal here is to provide consistent debug logging and error handling
+    // as we do in AxiosDataFetcher and GraphQLRequestClient
+    debug('request: %o', { url, ...requestInit });
+    const response = await fetchImpl(url, requestInit).catch((error) => {
+      debug('request error: %o', error);
+      throw error;
+    });
+
+    const isJson = response.headers.get('Content-Type')?.includes('application/json');
+    const respData = isJson ? await response.json() : undefined;
```
[the remark this answers is not in the document]
Comment (ambrauer) [reply packages/sitecore-jss/src/native-fetcher.ts:42]: Agree, I added a catch and debug for this case (no throw, response will just have `data` as `undefined`).

File: packages/sitecore-jss/src/native-fetcher.ts (hunk 6 above)
Comment (ambrauer) [reply packages/sitecore-jss/src/native-fetcher.ts:60]: I think these should be kept separate. In the other case, we are not throwing and also debug log should be a different message.
